src/spanned.rs

```diff
@@ -0,0 +1,153 @@
+//!
+//! #[derive(Deserialize)]
+//! struct Value {
+//!     s: Spanned<String>,
+//! }
+//!
+//! fn main() {
+//!     let t = "s = \"value\"\n";
+//!
+//!     let u: Value = toml::from_str(t).unwrap();
+//!
+//!     assert_eq!(u.s.start, 4);
+//!     assert_eq!(u.s.end, 11);
+//! }
+//! ```
+
+use serde::{de, ser};
+use std::fmt;
+
+#[doc(hidden)]
+pub const NAME: &'static str = "$__toml_private_Spanned";
+#[doc(hidden)]
+pub const START: &'static str = "$__toml_private_start";
+#[doc(hidden)]
+pub const END: &'static str = "$__toml_private_end";
+#[doc(hidden)]
+pub const VALUE: &'static str = "$__toml_private_value";
+
+macro_rules! key_deserialize {
+    ($ident:ident, $field:expr, $name:expr) => {
+        struct $ident;
+
+        impl<'de> de::Deserialize<'de> for $ident {
+            fn deserialize<D>(deserializer: D) -> Result<$ident, D::Error>
+                where D: de::Deserializer<'de>
+            {
+                struct FieldVisitor;
+
+                impl<'de> de::Visitor<'de> for FieldVisitor {
+                    type Value = ();
+
+                    fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+                        formatter.write_str("a valid spanned field")
+                    }
+
+                    fn visit_str<E>(self, s: &str) -> Result<(), E>
+                        where E: de::Error
+                    {
+                        if s == $field {
+                            Ok(())
+                        } else {
+                            Err(de::Error::custom(
+                                concat!("expected spanned field `", $name, "`")))
+                        }
+                    }
+                }
+
+                deserializer.deserialize_identifier(FieldVisitor)?;
+                Ok($ident)
+            }
+        }
+    }
+}
+
+
+/// A spanned value, indicating the range at which it is defined in the source.
+#[derive(Debug, Hash, PartialEq, Eq, PartialOrd, Ord)]
+pub struct Spanned<T> {
+    /// The start range.
+    pub start: usize,
+    /// The end range (exclusive).
+    pub end: usize,
+    /// The spanned value.
+    pub value: T,
+}
+
+impl<'de, T> de::Deserialize<'de> for Spanned<T>
+    where T: de::Deserialize<'de>
+{
+    fn deserialize<D>(deserializer: D) -> Result<Spanned<T>, D::Error>
+        where D: de::Deserializer<'de>
+    {
+        struct SpannedVisitor<T>(::std::marker::PhantomData<T>);
+
+        impl<'de, T> de::Visitor<'de> for SpannedVisitor<T>
+            where T: de::Deserialize<'de>
+        {
+            type Value = Spanned<T>;
+
+            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+                formatter.write_str("a TOML spanned")
+            }
+
+            fn visit_map<V>(self, mut visitor: V) -> Result<Spanned<T>, V::Error>
+                where V: de::MapAccess<'de>
+            {
+                let start = visitor.next_key::<StartKey>()?;
+
+                if start.is_none() {
```

Comment:
I would recommend cutting out the `key_deserialize` macro and using:

```rust
if visitor.next_key()? != Some(START) {
```

Reply:
Fixed, thanks!